brayns/engine/geometry/Geometry.h

```diff
@@ -0,0 +1,209 @@
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#pragma once
+
+#include <brayns/common/Bounds.h>
+
+#include <ospray/ospray_cpp/Geometry.h>
+
+#include <cassert>
+#include <numeric>
+#include <string_view>
+#include <vector>
+
+namespace brayns
+{
+/**
+ * @brief The RenderableBoundsUpdater is a templated class that can be specialized to update bounds
+ * based on a trasnformation and a given specialized geometry
+ */
+template<typename T>
+class GeometryBoundsUpdater
+{
+public:
+    /**
+     * @brief Updates a Bounds object with the given geometry trasnformed by the given matrix
+     * @param renderableType The source renderable used to update the bounds
+     * @param transform The matrix by which to trasnform the geometry
+     * @param bounds The bounds to update
+     */
+    static void update(const T &renderableType, const Matrix4f &transform, Bounds &bounds)
+    {
+        (void)renderableType;
+        (void)transform;
+        (void)bounds;
+    }
+};
+
+/**
+ * @brief Utility class to deduce the Ospray geometry name for a given geometry type
+ */
+template<typename T>
+class OsprayGeometryName
+{
+public:
+    /**
+     * @brief Returns the Ospray name of the type
+     */
+    static const std::string &get()
+    {
+        throw std::runtime_error("Unhandled geometry type");
+    }
+};
+
+template<typename T>
+class InputGeometryChecker
+{
+public:
+    static void check(const std::vector<T> &primitives)
+    {
+        if (primitives.size() >= std::numeric_limits<uint32_t>::max())
+        {
+            throw std::invalid_argument("Maximum number of primitives is 2^32 - 1");
+        }
+    }
+};
+
+/**
+ * @brief Utility class to commit geometry-specific parameters
+ */
+template<typename T>
+class GeometryCommitter
+{
+public:
+    /**
+     * @brief Commits the type specific parameters
+     */
+    static void commit(const ospray::cpp::Geometry &osprayGeometry, const std::vector<T> &primitives)
+    {
+        (void)osprayGeometry;
+        (void)primitives;
+        throw std::runtime_error("Unhandled geometry type");
+    }
+};
+
+/**
+ * @brief The Geometry class is a templated class to easily manage Ospray geometry at a high level.
+ * It allows to add, retrieve and manipulate Geometry.
+ */
+template<typename T>
+class Geometry
+{
+public:
+    Geometry(T primitive)
+        : _osprayGeometry(OsprayGeometryName<T>::get())
+    {
+        _primitives.push_back(std::move(primitive));
+        InputGeometryChecker<T>::check(_primitives);
```

Comment:
You make the check after doing the push_back, so the operation will still occurs while an error happened...
Is it intended or was the check supposed to be before the operation ?

Reply:
It is intended. To simplify the API, the class only check on vector of primitives. Any error that arises from this check mean the program will terminate, so there is no waste.